website/versioned_docs/version-0.17.3/getting-started/tutorials/single-page-application.md

```diff
@@ -0,0 +1,1842 @@
+
+![](/img/tutorials/single-page-application/rust-wasm-yew-single-page-application-7.png)
+
+This works fine for shallow component hierarchies but when you have a deep component hierarchy (which is common in larger SPAs), you will need to pass this state through multiple layers of components (which might not have use for this prop) to reach the desired node. This is called "Prop Drilling".
+
+You can see that `cart_products` is now passed from `App` to `AddToCart` component via `ProductDetail` and `Home` even though they have no use for this state. Imagine the same scenario with components many layers deep.
+
+This is the problem the global state solves. here is how it would look like:
+
+![](/img/tutorials/single-page-application/rust-wasm-yew-single-page-application-8.png)
+
+Notice how there is a direct link between the components that need this state and the global state.
+
+Unfortunately, Yew doesn't seem to have a [good solution](https://github.com/yewstack/yew/issues/576) for this. Today, the recommended solution is to use `Agents` for broadcasting state changes via pubsub. Yet, this approach might be hard to reason around. The story around state management in Yew is far from definitive and solutions to this problem may appear in the future.
+Let's solve our problem by hoisting the state.
+
+## Hoisting State
+
+We will be refactoring our code by moving `cart_products` state to `App` and extracting `Navbar` and `AtcButton` as separate components:
+
+![](/img/tutorials/single-page-application/rust-wasm-yew-single-page-application-9.png)
+
+```rust
+// src/components/navbar.rs
+use crate::types::CartProduct;
+use yew::prelude::*;
+
+pub struct Navbar {
+    props: Props,
+}
+
+#[derive(Properties, Clone)]
+pub struct Props {
+    pub cart_products: Vec<CartProduct>,
+}
+
+impl Component for Navbar {
+    type Message = ();
+    type Properties = Props;
+
+    fn create(props: Self::Properties, _link: ComponentLink<Self>) -> Self {
+        Self { props }
+    }
+
+    fn update(&mut self, _msg: Self::Message) -> ShouldRender {
+        true
+    }
+
+    fn change(&mut self, props: Self::Properties) -> ShouldRender {
+        self.props = props;
+        true
+    }
+
+    fn view(&self) -> Html {
+        let cart_value = self
+            .props
+            .cart_products
+            .iter()
+            .fold(0.0, |acc, cp| acc + (cp.quantity as f64 * cp.product.price));
+
+        html! {
+            <div class="navbar">
+                <div class="navbar_title">{"RustMart"}</div>
+              <div class="navbar_cart_value">{format!("${:.2}", cart_value)}</div>
+            </div>
+        }
+    }
+}
+```
+
+Notice how we started using the `change` lifecycle methods in the `Navbar` component. When the props sent from parent changes, we need to update the props inside the component so the UI re-renders.

```

Comment:
```suggestion
Notice how we started using the `change` lifecycle methods in the `Navbar` component. When the props sent from the parent change, we need to update the props inside the component so that the UI re-renders.

```